### ISSUE TITLE
How do I integrate "whatwasit" with "ex_admin"?

### ISSUE THREAD

**Issue by vraravam** [smpallen99/ex_admin#241]
I'm trying to incorporate the [whatwasit](https://github.com/smpallen99/whatwasit) package into the contact_demo project that uses ex_admin. I'm unable to follow since what I need is:
1.  Either the ex_admin should allow for an extra hook where the current user can be hooked into the changeset; (or)
2.  For the "changeset" method to be invoked with the 3rd (extra) optional parameter.

Any ideas on accomplishing this?

**Comment by vraravam** [comment]
@smpallen99 : Could you please help?
I have also asked the same question with some more details here: https://stackoverflow.com/questions/40927121/integrating-whatwasit-and-ex-admin

**Comment by smpallen99** [comment]
@vraravam You opened a hornets nest with this issue. I can't see an easy way for this integration at this time, especially with the current design for handling many to many relationships. @gwincr11 is working on refactoring the `ExAdmin.Changeset` and `ExAdmin.Repo` modules to make more use of `cast_assoc` in the model's changeset. After this refactor, I think it will be much easier to add support for this.

**Comment by gwincr11** [comment]
@vraravam yes, as @smpallen99 said I don't think you could do this today. I am looking at refactoring the controllers to use your models changeset instead of the ex_admin change set. When this was written, ecto did not have support for nested changesets so we needed some magic to break up has many writes. 

I think I have a pretty good idea how I am going to approach the refactor, I am hoping to get some tests in place in a pull request then start a separate request to get the refactor going.

**Comment by vraravam** [comment]
Not sure if this will help, but I got this working with a bit of a hack:
```
# Inside the register_resource block
    controller do
      before_filter :capture_audit_settings, only: [:create, :update, :delete]

      def capture_audit_settings(conn, params), do: Auditable.capture_audit_settings(conn, params)
    end

    changesets create: &Auditable.audited_changeset/2,
               update: &Auditable.audited_changeset/2,
               delete: &Auditable.audited_changeset/2
```
```
# lib\contact_demo\auditable.ex
defmodule ContactDemo.Auditable do
  alias ExAdmin.Utils
  alias ContactDemo.Authentication, as: Auth

  def capture_audit_settings(conn, params) do
    user = Auth.current_user(conn)
    # TODO: Not sure if injecting these into the params object is hacky or not
    new_params = params
      |> put_in([conn.assigns.defn.resource_name, :whodoneit_id], user.id)
      |> put_in([conn.assigns.defn.resource_name, :whodoneit_name], user.name)

    {conn, new_params}
  end

  def audited_changeset(model, params) do
    resource_name = Utils.resource_name(model.__struct__)
    resource_params = if Map.has_key?(params, resource_name), do: params[resource_name], else: params
    resource_params = resource_params || %{}
    {whodoneit_id, _} = Map.pop(resource_params, "whodoneit_id")
    {whodoneit_name, _} = Map.pop(resource_params, "whodoneit_name")
    model.__struct__.changeset(model, params, [whodoneit_id: whodoneit_id, whodoneit_name: whodoneit_name])
  end
end
```
Of course, this only handles the current object, and not any relationships (for now)...

**Comment by smpallen99** [comment]
Nice solution. Perhaps a little hackie, but should work until we get an official integration.
